stable/kommander-thanos/templates/query-stores-configmap.yaml

```diff
@@ -0,0 +1,15 @@
+{{ $labels := include "kommander-thanos.labels" . | indent 4  }}
+{{- if .Values.thanos.query.enabled }}
+{{- range .Values.thanos.query.serviceDiscoveryFileConfigMaps }}
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ . }}
+  labels:
+{{ $labels }}
+data:
+  stores.yaml: |-
+    - targets: []
```

Comment:
This will overwrite the targets on upgrade. Will that be updated by the controller? Alternatively, could we leave this configmap out and let the controller create it if necessary? This would mean thanos wouldn't start up until the configmap existed (created by controller on startup) but I think that's ok. What do you think?

Reply:
yeah, we should test an upgrade here, it could wipe out the stores :S

Reply:
I don't think the controller would be able to recover from losing all the configured stores in its current form, since it only acts on changes to individual `KommanderCluster` resources. For that to work, we'd need to be able to reconcile changes to this configmap by querying all existing `KommanderCluster`s, retrieving their alertmanager service endpoints, and then overwriting1 the configmap with those endpoints. Would that be a sensible thing to implement?

The alternative of giving the controller total ownership of the configmap's lifecycle should work as well. I think we'd want to expose an option in the KCL chart to define that configmap's name, so we don't hardcode it, and document that this chart will fail to deploy if the configmaps in `serviceDiscoveryFileConfigMaps` don't exist. But that seems reasonable.

Reply:
The only issue I found with giving the controller ownership over the configmap is that due to thanos failing to start up, addons deployment times out (`Error: failed to deploy the cluster: deployment timed out! the following addons were not resolved: kommander`), which isn't great for user experience.

Adding a pre-install hook to the configmap should avoid overwriting on upgrades. I tested this in a cluster by starting a cluster on a version of kommander-thanos, manually modifying the configmap, then redeploying addons with an updated version of kommander which points to a bumped version+1 of kommander-thanos, and seeing that the configmap is not overwritten upon upgrade. I got the idea from [this thread](https://github.com/helm/charts/issues/5167#issuecomment-388833226), only caveat is that the cm may not get deleted upon uninstallation of the chart, but I think that's fine.

This solution would not require any changes to the [KCL refactor PR](https://github.com/mesosphere/kommander-cluster-lifecycle/pull/231). Lmk what you think @hectorj2f @jimmidyson

Reply:
If you're adding a pre-install hook, you can add a job to delete it in a post-remove hook perhaps? I'm OK with that, but I'd really rather not leave anything lying around after uninstall.

Reply:
@jimmidyson @hectorj2f I added a post-delete hook job, take a look and let me know what you think, thanks!